### PR TITLE
Enable kw mail file and directory sending

### PR DESCRIPTION
**Pull request by JGBSouza**
solve issue #670

### CODE REVIEW

File: src/mail.sh [hunk 1]
```diff
@@ -186,8 +188,21 @@ function pre_generate_patches()
   mkdir -p "$patch_cache"
 
   cmd_manager 'SILENT' "git format-patch --quiet --output-directory $patch_cache $version $commit_range"
+}
```
Comment (rodrigosiqueira) [inline src/mail.sh:198]: It looks like you add an extra space here.

File: src/mail.sh [hunk 2]
```diff
@@ -99,6 +99,7 @@ function mail_send()
   local kernel_root
   local patch_count=0
   local cmd='git send-email'
+  declare -g patch_cache="${KW_CACHE_DIR}/patches"
```
Comment (rodrigosiqueira) [inline src/mail.sh:103]: Tbh, I don't see the value of making this variable global and in this function. If you think we can reuse it in multiple areas and we do not change it, declare it at the beginning of the file using upper case and the read-only option. Making this variable global inside this function is confusing.

File: tests/unit/mail_test.sh [hunk 3]
```diff
@@ -9,7 +9,9 @@ function oneTimeSetUp()
   declare -gr FAKE_GIT="$SHUNIT_TMPDIR/fake_git/"
   declare -gr FAKE_KERNEL="$FAKE_GIT/fake_kernel/"
   declare -ga test_config_opts=('test0' 'test1' 'test2' 'user.name' 'sendemail.smtpuser')
-
+  declare -gr FAKE_PATCHES_DIR="fake_patches"
```
Comment (rodrigosiqueira) [inline tests/unit/mail_test.sh:12]: single quotes

File: tests/unit/mail_test.sh [hunk 4]
```diff
@@ -30,6 +32,14 @@ function oneTimeSetUp()
 
   mk_fake_git
 
+  mkdir "./$FAKE_PATCHES_DIR"
```
Comment (rodrigosiqueira) [inline tests/unit/mail_test.sh:36]: https://kworkflow.org/content/codingstyle.html#string-concatenation

File: tests/unit/mail_test.sh [hunk 5]
```diff
@@ -30,6 +32,14 @@ function oneTimeSetUp()
...
+  mkdir "./$FIRST_TEST_PATCHES_DIR"
+  cp -f "${ORIGINAL_DIR}/tests/unit/samples/test.patch" "${FIRST_TEST_PATCHES_DIR}/first.patch"
```
Comment (rodrigosiqueira) [inline tests/unit/mail_test.sh:39]: https://kworkflow.org/content/codingstyle.html#avoid-short-options

File: tests/unit/mail_test.sh [hunk 6]
```diff
@@ -204,6 +214,32 @@ function test_validate_email_list()
...
+function test_count_patches()
+{
+  cd "$FAKE_GIT" || {
+    ret="$?"
+    fail "($LINENO): Failed to move to fake git dir"
+    exit "$ret"
+  }
+
+  local output
+  local expected
```
Comment (rodrigosiqueira) [inline tests/unit/mail_test.sh:228]: move this to the beginning of this function.

File: tests/unit/mail_test.sh [hunk 7]
```diff
@@ -204,6 +214,32 @@ function test_validate_email_list()
   assert_equals_helper 'Expected a success' "$LINENO" "$ret" 0
 }
 
+function test_count_patches()
+{
+  cd "$FAKE_GIT" || {
+    ret="$?"
```
Comment (rodrigosiqueira) [inline tests/unit/mail_test.sh:222]: declare `ret` as local variable.

File: tests/unit/mail_test.sh [hunk 8]
```diff
@@ -204,6 +214,32 @@ function test_validate_email_list()
...
+  local output
+  local expected
+
+  output="$(count_patches "$FIRST_TEST_PATCHES_DIR")"
+  expected=2
+  assert_equals_helper 'Expected two patches' "$LINENO" "$output" "$expected"
+
+  output="$(count_patches "$SECOND_TEST_PATCHES_DIR")"
+  expected=1
+  assert_equals_helper 'Expected two patches' "$LINENO" "$output" "$expected"
```
Comment (rodrigosiqueira) [inline tests/unit/mail_test.sh:236]: You should add other tests that validate:
1. There is not patch.
2. The target file is not a patch.

File: src/mail.sh [hunk 9]
```diff
@@ -115,8 +116,14 @@ function mail_send()
     cmd+=" --cc=\"$cc_recipients\""
   fi
 
-  pre_generate_patches "$commit_range" "$version"
-  patch_count="$(count_patches "$patch_cache")"
+  if [[ -z "$patch_source" ]]; then
+    pre_generate_patches "$commit_range" "$version"
+    patch_count="$(count_patches "$patch_cache")"
```
Comment (rodrigosiqueira) [inline src/mail.sh:121]: Did you validate if your change does not break the original behavior? I mean, the one using the commit SHA.

File: src/mail.sh [hunk 10]
```diff
@@ -115,8 +116,14 @@ function mail_send()
...
+  elif [[ -d "$patch_source" ]]; then
+    patch_count="$(count_patches "$patch_source")"
+  elif [[ $(is_a_patch "$patch_source") ]]; then
+    patch_count=1
+  fi
```
Comment (rodrigosiqueira) [inline src/mail.sh:126]: Add a comment inside each if that highlight the case the if condition is handling.

File: src/mail.sh [hunk 11]
```diff
@@ -135,7 +142,7 @@ function mail_send()
   [[ -n "$private" ]] && cmd+=" $private"
   [[ -n "$rfc" ]] && cmd+=" $rfc"
   [[ -n "$extra_opts" ]] && cmd+=" $extra_opts"
-
+  [[ -n "$patch_source" ]] && cmd+=" $patch_source"
```
Comment (rodrigosiqueira) [inline src/mail.sh:145]: https://kworkflow.org/content/codingstyle.html#string-concatenation

File: src/mail.sh [hunk 12]
```diff
@@ -1119,6 +1126,8 @@ function parse_mail_options()
         for i in "$@"; do
           if [[ "${i}" =~ " " ]]; then
             pass_option_to_send_email+=" ${i@Q}"
+          elif [[ -f "$i" || -d "$i" ]]; then
+            options_values['PATCHES_SOURCE']="$i"
```
Comment (rodrigosiqueira) [inline src/mail.sh:1130]: The interface is not clear to me. You must update the man page by adding how this new option looks like.

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]

**Comment by davidbtadokoro** [comment]
Hi @JGBSouza, and thank you so much for your change!

I spent maybe an hour+ reviewing your PR and thinking if it was the best approach and testing it, but it suddenly hit me that maybe we don't need to do all the handling of the number of patches, and the processing of files/directories which can represent a patchset.

Below, are my takes on the matter, which you can consider my "review" (@rodrigosiqueira, if possible, also give me your thoughts on the matter).

# My interpretation of the problem:

Look at what Rubens described in the issue:

> Currently the biggest problem lies with the automatic inclusion of the reference to @^ when no revision ranges are passed.

This is in reference to [this part](https://github.com/kworkflow/kworkflow/blob/af4d2e3100bd33c8dd07c8dba2c42cd7af020c9b/src/mail.sh#L1130) of kw mail, which means that, if we don't pass any commit reference (either a range, or a number of commits) to `--send`, the value of `options_values['COMMIT_RANGE']` would be `@^`, which, FWIU, is the same as `HEAD^`, meaning the last commit. What he is alluding to is the following:

Take this repo that I have in my machine with this git log:

2024-04-28-125222_509x189_scrot

I create the following patch files that represent the patchset with

2024-04-28-125343_987x279_scrot

I can then use kw mail to send the patchset (almost) correctly with the current unstable `kw mail` implementation (notice that `kw mail --send -- *.patch` would work the same)

2024-04-28-125635_744x353_scrot

The problem is that the last commit (represented by file `/tmp/4kZTrnWu7j/0001-Add-Sofware.patch`) is also included in the sending of the patch.

# How `--send` really works:

It is really important to notice that `--send` actually doesn't receive a argument, per se. Look at the [case](https://github.com/kworkflow/kworkflow/blob/af4d2e3100bd33c8dd07c8dba2c42cd7af020c9b/src/mail.sh#L1000) in the parser and see that `--send` is treated as a standalone option for `kw mail`, so values like `-3` or `abcd..e1d3` are treated as additional arguments (which fall in [this case](https://github.com/kworkflow/kworkflow/blob/af4d2e3100bd33c8dd07c8dba2c42cd7af020c9b/src/mail.sh#L1097) of the parser).

This means that commit ranges are directly processed by `git send-email`, which is good as delegating this to git reduces the complexity and problems to us.

# My opinion on the best approach:

Pay attention that `kw mail` currently supports the passing of files/directories with patches like Rubens described, but it sends the last commit together, which is wrong.

I think the best approach would be to modify the function `find_commit_references` (see [this line](https://github.com/kworkflow/kworkflow/blob/af4d2e3100bd33c8dd07c8dba2c42cd7af020c9b/src/mail.sh#L1113) for reference) to not return 22 EINVAL when files or directories are passed as arguments.

In other words, IMHO, the best approach is to expand the current behavior to just **also** support the passing of files and directories. If I am correct, with this approach the change in the codebase will be minimal and we would also benefit from the complexity to handle the processing of files and directories being totally delegated to `git send-email`.

**Comment by davidbtadokoro** [comment]
# UPDATE

@JGBSouza, I've managed to implement the necessary support for files and directories with the following

```
diff --git a/src/mail.sh b/src/mail.sh
index f4a6981..d5ca143 100644
--- a/src/mail.sh
+++ b/src/mail.sh
@@ -347,6 +347,10 @@ function find_commit_references()
     return 0
   fi
 
+  for arg in $args; do
+    [[ -f "$arg" || -d "$arg" ]] && return 0
+  done
+
   return 22 # EINVAL
 }
```

As always, code and approach suggestions from my end are to be taken with a grain of salt. In other words, you need to thoroughly and actively think on my arguments and don't take anything I say for granted. If you feel like my approach is best, heavily validate, refine and test it!

File: src/mail.sh (hunk 2 above)
Comment (JGBSouza) [reply src/mail.sh:103]: I also think that this is the case of making the variable global in the beginnig of the file, but for some reason, "$KW_CACHE_DIR" is an empty var in the begining of the code and i didn't understand where it's value is updated inside this code

File: src/mail.sh [hunk 13]
```diff
@@ -342,6 +342,10 @@ function find_commit_references()
     parsed=''
   done <<< "$(git rev-parse -- $args 2> /dev/null)"
 
+  if [[ -f "$i" || -d "$i" ]]; then
+    return 0
+  fi
```
Comment (davidbtadokoro) [inline src/mail.sh:347]: I am a little confused, where does `i` come from? I don't see it declared anywhere in this function neither see any relation with `args`, which is the variable that stores the files/directory arguments. It seems to work, as `kw mail --send` and `kw mail --send *.patch` work as intended (i.e., the first sends the last commit as a patch, and the second sends the patch files as patches **without** also including the last commit), but it isn't clear how. Please, comment your reasoning behind this, in case I am mistaken about it being wrong.

There are also two points:

1. As `kw mail` "enhances" `git send-email` we have an opportunity to check if the arguments are valid patch files. So I think we could do some kind of loop to check if each argument is a patch file (in case they are files) or a directory containing patch files (in case it is a directory).
2. I think the check to see if the arguments are patch files/directories should come after the below check (`if [[ -n "$commit_range" ]]; then`), as this last one is related with the while loop above (`while read -r arg; do`). I am saying this, because this function now finds commit references **or** patch files/directories, so it should check those separately.

File: src/mail.sh
Comment (davidbtadokoro) [inline src/mail.sh:1]: 1. You should probably rename `find_commit_references` as now it checks for commit references **or** patch files/dirs, Find a name that transmits both, in a more succinct way than `find_commit_references_or_patch_files_dirs`, if possible.
2. You should modify the function documentation to reflect this change in behavior.

File: tests/unit/mail_test.sh [hunk 14]
```diff
@@ -10,6 +10,9 @@ function oneTimeSetUp()
   declare -gr FAKE_KERNEL="$FAKE_GIT/fake_kernel/"
   declare -ga test_config_opts=('test0' 'test1' 'test2' 'user.name' 'sendemail.smtpuser')
 
+  declare -gr FAKE_PATCHES_DIR='fake_patches'
+  declare -gr FAKE_PATCH_FILE="${FAKE_PATCHES_DIR}/fake_patch_file"
```
Comment (davidbtadokoro) [inline tests/unit/mail_test.sh:14]: I know that because of the `cd` statement below, we will eventually be at `FAKE_GIT`, and it will be safe to do what you did:

```bash
mkdir "./${FAKE_PATCHES_DIR}"
touch "./${FAKE_PATCH_FILE}"
```

However, I think the cleanest way is to define explicitly as `FAKE_PATCHES_DIR=${FAKE_GIT}/fake_patches` and make the statements below simply:

```bash
  mkdir "$FAKE_PATCHES_DIR"
  touch "$FAKE_PATCH_FILE"
```

this way we don't depend on relative paths which can induce errors that will be hard to debug.

File: tests/unit/mail_test.sh [hunk 15]
```diff
@@ -30,6 +33,9 @@ function oneTimeSetUp()
 
   mk_fake_git
 
+  mkdir "./${FAKE_PATCHES_DIR}"
+  touch "./${FAKE_PATCH_FILE}"
```
Comment (davidbtadokoro) [inline tests/unit/mail_test.sh:37]: See the comment above to change these lines accordingly

File: tests/unit/mail_test.sh [hunk 16]
```diff
@@ -513,6 +529,7 @@ function test_mail_send()
     fail "($LINENO): Failed to move back to original dir"
     exit "$ret"
   }
+
```
Comment (davidbtadokoro) [inline tests/unit/mail_test.sh:532]: You don't need to add this line, remove it

File: tests/unit/mail_test.sh [hunk 17]
```diff
@@ -494,6 +500,16 @@ function test_mail_send()
   expected='git send-email --to="mail@test.com" extra_args --other_arg -13 -v2'
   assert_equals_helper 'Testing no options option' "$LINENO" "$expected" "$output"
 
+  parse_mail_options "$FAKE_PATCH_FILE" '--to=mail@test.com'
+  output=$(mail_send 'TEST_MODE')
+  expected="git send-email --to=\"mail@test.com\" $FAKE_PATCH_FILE"
```
Comment (davidbtadokoro) [inline tests/unit/mail_test.sh:505]: String concatenation in `FAKE_PATCH_FILE`

File: tests/unit/mail_test.sh [hunk 18]
```diff
@@ -494,6 +500,16 @@ function test_mail_send()
...
+  assert_equals_helper 'Testing send with a file' "$LINENO" "$expected" "$output"
+
+  parse_mail_options "$FAKE_PATCHES_DIR" '--to=mail@test.com'
+  output=$(mail_send 'TEST_MODE')
+  expected="git send-email --to=\"mail@test.com\" $FAKE_PATCHES_DIR"
```
Comment (davidbtadokoro) [inline tests/unit/mail_test.sh:510]: String concatenation in `FAKE_PATCHES_DIR`

File: tests/unit/mail_test.sh [hunk 19]
```diff
@@ -494,6 +500,16 @@ function test_mail_send()
   expected='git send-email --to="mail@test.com" extra_args --other_arg -13 -v2'
   assert_equals_helper 'Testing no options option' "$LINENO" "$expected" "$output"
 
+  parse_mail_options "$FAKE_PATCH_FILE" '--to=mail@test.com'
```
Comment (davidbtadokoro) [inline tests/unit/mail_test.sh:503]: Also, I would like to see more real life cases:

1. More than one patch file without cover letter
2. More than one patch file with cover letter
3. Use of patch files with prefix `000` (as in `0003` or `0010`) and suffix `.patch` to relate to what `git format-patches` outputs

File: src/mail.sh
Comment (davidbtadokoro) [inline src/mail.sh:1]: Add an example in `kw mail` manpage to exemplify the passing of patch files/directories as arguments.

**Review by davidbtadokoro — CHANGES_REQUESTED** [review]
Hi @JGBSouza, and thank you for the update!

I really think this approach is a cleaner and direct way to implement the desired behavior, without the risk of adding any regression.

I left important comments so please tackle everyone of them (ping in case of **any** doubt), but mind that this PR is looking legit. It works as intended, so now we should focus on refining it.

**Update:** Forgot to mention that you didn't add a `Signed-off-by` line in the commit message, so please do it.

**Comment by rodrigosiqueira** [comment]
@JGBSouza any updates here?

**Comment by rodrigosiqueira** [comment]
PR closed due to the lack of updates.
